File: backend/app/services/brain_service.py
"""Brain CRUD service."""

import uuid

from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import AuraException
from app.models.brain import Brain
from app.models.document import Document
from app.repositories.brain_repository import BrainRepository
from app.repositories.document_repository import DocumentRepository
from app.schemas.brain import BrainCreate, BrainUpdate


class BrainService:
    """Handles brain creation, updates, deletion, and document association."""

    def __init__(self, db: AsyncSession) -> None:
        self._repo = BrainRepository(db)
        self._doc_repo = DocumentRepository(db)

    async def create(self, user_id: uuid.UUID, payload: BrainCreate) -> Brain:
        """Create a new brain owned by the given user."""
        return await self._repo.create(user_id, payload)

    async def list_for_user(self, user_id: uuid.UUID) -> list[Brain]:
        """Return all brains belonging to the user."""
        return await self._repo.list_by_user(user_id)

    async def get(self, brain_id: uuid.UUID, user_id: uuid.UUID) -> Brain:
        """Fetch a brain, ensuring it belongs to the requesting user."""
        brain = await self._repo.get_by_id(brain_id)
        if not brain or brain.user_id != user_id:
            raise AuraException(detail="Brain not found.", code="BRAIN_NOT_FOUND")
        return brain

    async def update(self, brain_id: uuid.UUID, user_id: uuid.UUID, payload: BrainUpdate) -> Brain:
        """Update a brain's metadata."""
        brain = await self.get(brain_id, user_id)
        return await self._repo.update(brain, payload)

    async def delete(self, brain_id: uuid.UUID, user_id: uuid.UUID) -> None:
        """Delete a brain."""
        brain = await self.get(brain_id, user_id)
        await self._repo.delete(brain)

    async def list_documents(self, brain_id: uuid.UUID, user_id: uuid.UUID) -> list[Document]:
        """Return documents connected to a brain."""
        await self.get(brain_id, user_id)  # authorization check
        return await self._repo.list_documents(brain_id)

    async def add_document(
        self, brain_id: uuid.UUID, document_id: uuid.UUID, user_id: uuid.UUID
    ) -> None:
        """Attach a document to a brain."""
        await self.get(brain_id, user_id)
        doc = await self._doc_repo.get_by_id(document_id)
        if not doc:
            raise AuraException(detail="Document not found.", code="DOCUMENT_NOT_FOUND")
        await self._repo.add_document(brain_id, document_id)

    async def remove_document(
        self, brain_id: uuid.UUID, document_id: uuid.UUID, user_id: uuid.UUID
    ) -> None:
        """Detach a document from a brain."""
        await self.get(brain_id, user_id)
        await self._repo.remove_document(brain_id, document_id)

File: frontend/src/services/brain-service.ts
/**
 * Brain service — CRUD and document association calls to FastAPI.
 */
import { apiClient } from "@/lib/api-client";
import type { Brain, BrainCreate, BrainDocument, BrainUpdate } from "@/types";

export const brainService = {
  /** List all brains for the authenticated user. */
  async list(): Promise<Brain[]> {
    const { data } = await apiClient.get<Brain[]>("/brains");
    return data;
  },

  /** Create a new brain. */
  async create(payload: BrainCreate): Promise<Brain> {
    const { data } = await apiClient.post<Brain>("/brains", payload);
    return data;
  },

  /** Update a brain's name or description. */
  async update(brainId: string, payload: BrainUpdate): Promise<Brain> {
    const { data } = await apiClient.patch<Brain>(`/brains/${brainId}`, payload);
    return data;
  },

  /** Delete a brain. */
  async remove(brainId: string): Promise<void> {
    await apiClient.delete(`/brains/${brainId}`);
  },

  /** List documents connected to a brain. */
  async listDocuments(brainId: string): Promise<BrainDocument[]> {
    const { data } = await apiClient.get<BrainDocument[]>(`/brains/${brainId}/documents`);
    return data;
  },

  /** Attach a document to a brain. */
  async addDocument(brainId: string, documentId: string): Promise<void> {
    await apiClient.post(`/brains/${brainId}/documents/${documentId}`);
  },

  /** Detach a document from a brain. */
  async removeDocument(brainId: string, documentId: string): Promise<void> {
    await apiClient.delete(`/brains/${brainId}/documents/${documentId}`);
  },
};
